### PR TITLE
Extract Sawtooth XO Test to an Example

### PR DESCRIPTION
**Pull request by swetharepakula**
- the example uses the XO transaction family to play a game of tic tac
 toe.
 - the example shows how to  use trasact end to end with multiple batches
 with state changes to the database in between

Signed-off-by: Swetha Repakula <srepaku@us.ibm.com>

### CODE REVIEW

File: examples/simple_xo/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,246 @@
+use sawtooth_xo::handler::XoTransactionHandler;
```
Comment (vaporos) [inline examples/simple_xo/src/main.rs]: Please add copyright headers to all the files (following the other files in the repo). Don't just blindly copy the 'Copyright' line though, make sure it's accurate. If you feel this is sufficiently derived from other code, include multiple 'Copyright' lines (the one(s) from the other code plus one for (presumably) IBM).

File: examples/simple_xo/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "simple_xo"
+version = "0.1.0"
+authors = ["Swetha Repakula <srepaku@us.ibm.com>"]
```
Comment (vaporos) [inline examples/simple_xo/Cargo.toml]: Make this match the copyright holder(s). This is an active topic of discussion, as to what we should put here, but that's the approach used in the other Cargo.toml file currently.

File: examples/simple_xo/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "simple_xo"
```
Comment (vaporos) [inline examples/simple_xo/Cargo.toml]: Both _ and - are valid here, but we've been preferring '-' based on what other "good" projects chose to do.

File: examples/simple_xo/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    }
+}
+
+fn print_current_state(value: &Vec<u8>) -> &str {
+    let val = match str::from_utf8(&value) {
+        Ok(v) => v,
+        Err(e) => panic!("Invalid UTF-8 sequence: {}", e),
+    };
+
+    let split: Vec<&str> = val.split(",").collect();
```
Comment (vaporos) [inline examples/simple_xo/src/main.rs]: A future enhancement (maybe after this PR?) might be create a Board struct and implement Display for it, and then do Board::from(&str) and println!("{}", board). Same with Game actually.

File: examples/simple_xo/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+use transact::signing::{hash::HashSigner, Signer};
+use transact::state::merkle::{self, MerkleRadixTree, MerkleState};
+use transact::state::StateChange as ChangeSet;
+use transact::state::Write;
+
+use hex;
+use sha2::{Digest, Sha512};
+use std::io;
+use std::str;
+use std::sync::{Arc, Mutex};
```
Comment (vaporos) [inline examples/simple_xo/src/main.rs]: We should be able to write this example without Arc/Mutex. But, I am also comfortable with this version going in and then iterating on it as an overall team.

File: examples/simple_xo/Cargo.toml (hunk 3 above)
Comment (swetharepakula) [reply examples/simple_xo/Cargo.toml]: Should I do the same with the directory name?

File: examples/simple_xo/src/main.rs (hunk 4 above)
Comment (swetharepakula) [reply examples/simple_xo/src/main.rs]: I like that! I will work on that

File: examples/simple_xo/Cargo.toml (hunk 3 above)
Comment (vaporos) [reply examples/simple_xo/Cargo.toml]: No, everywhere else _ is appropriate.

**Comment by vaporos** [comment]
Tested this. Very cool to have this be an interactive example.

Currently, it is very easy to cause an error which will make the demo exit. We may want to put in a feature in a subsequent PR that prevents invalid transactions from being fatal.

**Comment by vaporos** [comment]
Needs updates so that the build passes (currently failing on clippy errors).

File: examples/simple_xo/src/main.rs (hunk 5 above)
Comment (peterschwarz) [reply examples/simple_xo/src/main.rs]: I can see why this was done (based on the tests), but it definitely works without.

**Comment by swetharepakula** [comment]
I am not sure how to fix this error: https://build.sawtooth.me/job/Transact-Hyperledger/job/transact/job/PR-20/6/execution/node/23/log/

**Comment by rberg2** [comment]
> I am not sure how to fix this error: https://build.sawtooth.me/job/Transact-Hyperledger/job/transact/job/PR-20/6/execution/node/23/log/

Hello, this just means that a maintainer will need to start the build for you. I have started the build now. I apologize for not noticing and starting this sooner. Feel free to ping my on rocket chat if this happens again.

**Comment by swetharepakula** [comment]
Thanks!

File: examples/simple_xo/src/main.rs (hunk 5 above)
Comment (arsulegai) [reply examples/simple_xo/src/main.rs]: Because there's one instance of executor and the execution happen serially?

**Comment by swetharepakula** [comment]
Hi, what are the remaining things left for this PR to be merged in?

**Pull request merged**